### DISCUSSION TITLE
Anonymously aliased model is omitted from mappings()

### DISCUSSION THREAD

**Discussion by kdwyer** [sqlalchemy/sqlalchemy#10957] [Usage Questions] [answered]
Description:
When a ORM model class is anonymously aliased and queried, keys and values for the aliased model do not appear in the output of `mappings()` I would expect the behaviour to be consistent with models which are not aliased, or aliased with names, that is the mappings should contain keys with the model (or alias) name and the selected values.

SQLAlchemy version: 2.0.25
DB-API: stdlib sqlite3
DB: SQLite
Python version: 3.12.0
O/S Linux

Repro:
```python
import sqlalchemy as sa
from sqlalchemy import orm
from sqlalchemy.orm import Mapped, mapped_column


class Base(orm.DeclarativeBase):
    pass


class Foo(Base):
    __tablename__ = 'foos'

    id: Mapped[int] = mapped_column(primary_key=True)


engine = sa.create_engine('sqlite://', echo=True)
Base.metadata.create_all(engine)

with orm.Session(engine) as s:
    s.add(Foo())

    q0 = sa.select(Foo)
    r0 = s.execute(q0).mappings().all()
    print(f'{r0=}')

    a1 = orm.aliased(Foo, name='Bar')
    q1 = sa.select(a1)
    r1 = s.execute(q1).mappings().all()
    print(f'{r1=}')

    a2 = orm.aliased(Foo)
    q2 = sa.select(a2)
    r2a = s.execute(q2).all()
    print(f'{r2a=}')
    r2 = s.execute(q2).mappings().all()
    print(f'{r2=}')
```

Output:
```none
<snip>
2024-02-02 19:29:52,217 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-02-02 19:29:52,218 INFO sqlalchemy.engine.Engine INSERT INTO foos DEFAULT VALUES
2024-02-02 19:29:52,218 INFO sqlalchemy.engine.Engine [generated in 0.00010s] ()
2024-02-02 19:29:52,219 INFO sqlalchemy.engine.Engine SELECT foos.id 
FROM foos
2024-02-02 19:29:52,219 INFO sqlalchemy.engine.Engine [generated in 0.00010s] ()
r0=[{'Foo': <__main__.Foo object at 0x7ffa108550d0>}]
2024-02-02 19:29:52,220 INFO sqlalchemy.engine.Engine SELECT "Bar".id 
FROM foos AS "Bar"
2024-02-02 19:29:52,220 INFO sqlalchemy.engine.Engine [generated in 0.00009s] ()
r1=[{'Bar': <__main__.Foo object at 0x7ffa108550d0>}]
2024-02-02 19:29:52,221 INFO sqlalchemy.engine.Engine SELECT foos_1.id 
FROM foos AS foos_1
2024-02-02 19:29:52,221 INFO sqlalchemy.engine.Engine [generated in 0.00010s] ()
r2a=[(<__main__.Foo object at 0x7ffa108550d0>,)]
2024-02-02 19:29:52,221 INFO sqlalchemy.engine.Engine SELECT foos_1.id 
FROM foos AS foos_1
2024-02-02 19:29:52,221 INFO sqlalchemy.engine.Engine [cached since 0.000423s ago] ()
r2=[{}]
2024-02-02 19:29:52,222 INFO sqlalchemy.engine.Engine ROLLBACK
```

I have no use case for this behaviour, so feel free to wontfix or docfix if you see fit.

**Comment by zzzeek** [comment]
> that is the mappings should contain keys with the model (or alias) name 

the alias doesn't have a name.  what would the keys be named?

**Reply by kdwyer** [reply]
I suppose the alias generated by `aliased` would be the best candidate - so "foos_1" in the example. Is the current behaviour by design?

**Reply by zzzeek** [reply, marked as answer]
> I suppose the alias generated by aliased would be the best candidate - so "foos_1" in the example

that generation is not at all deterministic.  it can be affected by:

1. the order of columns / tables in the columns clause
2. the order in which parts of the query are parsed, which can change:
    a. when the query is modified to have things like CTEs, joinedloads, different WHERE criteria
    b. when using different database dialects / backends
    c. with different versions of SQLAlchemy (even minor versions)
3. the dialect / backend overall, the specific version of the database in use
4. the version of SQLAlchemy in use (even minor versions)

Basically, it is absolutely a very bad idea to write source code that assumes anything about these names.   that source code will break, unpredictably.      

SQLAlchemy in the larger sense is philosophically against linking parts of code together based on naming conventions at all, even when naming conventions are controllable.    this is why when you map a class, you have to give it a table name if you expect there to be a table present.   We aren't going to just default to "class name with lower case" or something like that.
  
> Is the current behaviour by design?

very, very much so!  Older versions of SQLAlchemy used to spit out these names which were of course very non-deteministic and we hoped people weren't coding to them.  that these names know how to be ignored from namespaces since they are not namespace names was a very deliberate and carefully implemented behavior.

Python has a phrase "explicit is better than implicit".  if you are expecting to target parts of your row by name, then give those columns names, explicitly.

**Reply by kdwyer** [reply]
If the behaviour is by design, then I'm happy for it to stay that way - for all the reasons that you've listed. Cheers!
